railties/lib/rails/generators/actions.rb

```diff
@@ -268,14 +268,25 @@ def rails_command(command, options = {})
       #   route "root 'admin#index'", namespace: :admin
       def route(routing_code, namespace: nil)
         routing_code = Array(namespace).reverse.reduce(routing_code) do |code, ns|
-          "namespace :#{ns} do\n#{indent(code, 2)}\nend"
+          "namespace :#{ns} do\n#{optimize_indentation(code, 2)}end"
         end
 
         log :route, routing_code
-        sentinel = /\.routes\.draw do\s*\n/m
+
+        after_pattern = Array(namespace).each_with_index.reverse_each.reduce(nil) do |pattern, (ns, i)|
+          margin = "\\#{i + 1}[ ]{2}"
+          "(?:(?:^[ ]*\n|^#{margin}.*\n)*?^(#{margin})namespace :#{ns} do\n#{pattern})?"
+        end.then do |pattern|
+          /^([ ]*).+\.routes\.draw do[ ]*\n#{pattern}/
+        end
 
         in_root do
-          inject_into_file "config/routes.rb", optimize_indentation(routing_code, 2), after: sentinel, verbose: false, force: false
+          if existing = match_file("config/routes.rb", after_pattern)
+            base_indent, *, prev_indent = existing.captures.compact.map(&:length)
+            routing_code = optimize_indentation(routing_code, base_indent + 2).lines.grep_v(/^[ ]{,#{prev_indent}}\S/).join
+          end
+
+          inject_into_file "config/routes.rb", routing_code, after: after_pattern, verbose: false, force: false
```

Comment:
What do you think about making the routes file configurable since `draw` was added in #37892 ?

For example, with the admin gem I'm working on, it'd be great to be able to add `draw :admin` to the main routes.rb and have our generators add routes to `config/routes/admin.rb`

Reply:
That sounds like a good idea.  Though I would prefer to do it in a follow-up PR.